alvr/sockets/src/stream_socket/mod.rs

```diff
@@ -144,35 +145,41 @@ impl<T: Serialize> StreamSender<T> {
         };
     }
 
-    pub async fn send(&mut self, header: &T, buffer: Vec<u8>) -> StrResult {
+    pub async fn send(&mut self, header: &T, payload_buffer: Vec<u8>) -> StrResult {
         // packet layout:
         // [ 2B (stream ID) | 4B (packet index) | 4B (packet shard count) | 4B (shard index)]
         // this escluses length delimited coding, which is handled by the TCP backend
         const OFFSET: usize = 2 + 4 + 4 + 4;
-        let max_payload_size = self.max_packet_size - OFFSET;
+        let max_shard_data_size = self.max_packet_size - OFFSET;
 
-        let data_header_size = bincode::serialized_size(header).map_err(err!()).unwrap() as usize;
-
-        let shards = buffer.chunks(max_payload_size);
-        let shards_count = shards.len() + 1;
+        let header_size = bincode::serialized_size(header).map_err(err!()).unwrap() as usize;
+        let header_shards_count = if header_size == 0 {
+            0
+        } else {
+            (header_size - 1) / max_shard_data_size + 1 // ceil division
+        };
+        self.header_buffer.clear();
+        if self.header_buffer.capacity() < header_size {
+            // If the buffer is empty, with this call we request a capacity of "header_size".
+            self.header_buffer.reserve(header_size);
+        }
+        bincode::serialize_into(&mut self.header_buffer, header)
+            .map_err(err!())
+            .unwrap();
+        let header_shards = self.header_buffer.chunks(max_shard_data_size);
```

Comment:
I think you can use `header_shards.len()` to get the number of header shards, so you don't have to calculate them above.

Reply:
Yeah, good catch